mautic/mautic
# WIP Tokenize preview email page

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



[//]: # ( Required: )
#### Description:
As a user, I would like to be able to more easily preview emails & landing pages and their variations including:


 - Translations (if available)
 - A/B tests
 - Dynamic content & tokens - [ ] 

#### Steps to test this PR:
1. Create an email with at least one AB test and one translation
2. Use at leas in one template tokens like `{contactfield=firstname}‍ {contactfield=lastname}‍` to see if contact info is filled properly
3. Test translations and variants preview in email detail page :

![Snímek obrazovky 2021-02-01 v 20 39 21](https://user-images.githubusercontent.com/12815758/106509198-96711a00-64cd-11eb-8d1c-d648c434fe82.png)

4. Use the same scenario for pages

#### Steps to test permissions
Preview actions have now protected access for admins and other roles. Now it is not accessible by anonymous users. There's a need to test it under other roles than admin which has no defined resource access right. I've a `Sales` role for this.
1. Use an admin account and previews are accessible
2. Use `Sales` or other role defined in mautic with `Email permissions - View Own & View Others` and `Landing page permissions - View Own & View Others` - works with respect to these settings against the user with this role
3. Try to  access previews in the anonymous window - this will return access denied

#### TODO:
- [x] Preview permissions fix (it was open to the public)
 - [x] emails
 - [x] Translations (if available)
 - [x] A/B tests
 - [x] Dynamic content & tokens - [ ] 

 - [x] landing pages
 - [x] Translations (if available)
 - [x] A/B tests
 - [x] Translations of pages ~~not available in controller/action and view. There's bug in `\Mautic\CoreBundle\Entity\TranslationEntityTrait::getTranslations`~~ - this was my mistake
 - [x] Dynamic content & tokens - contact tokens are not supported for pages
 - [x] Bug in contact lookup field. When used contact is edited or deleted, nothing changes
 - [x] The contact lookup box and the contact data in the preview should appear only to users that have permission to access contacts.
 - [x] Respect to public preview switch
 - [x] Respect to contact view permissions in preview



<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11156"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
